packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +460,347 @@ class ShellRoute extends RouteBase {
+/// can instead be built using the child widgets of the branches
+/// (see [StatefulShellRouteState.children]) directly. For example:
+///
+/// ```
+/// final StatefulShellRouteState shellState = StatefulShellRoute.of(context);
+/// final int currentIndex = shellState.currentIndex;
+/// final List<Widget?> children = shellRouteState.children;
+/// return MyCustomShell(currentIndex, children);
+/// ```
+///
+/// Below is a simple example of how a router configuration with
+/// StatefulShellRoute could be achieved. In this example, a
+/// BottomNavigationBar with two tabs is used, and each of the tabs gets its
+/// own Navigator. A container widget responsible for managing the Navigators
+/// for all route branches will then be passed as the child argument
+/// of the builder function.
+///
+/// ```
+/// final GoRouter _router = GoRouter(
+///   initialLocation: '/a',
+///   routes: <RouteBase>[
+///     StatefulShellRoute(
+///       routes: <RouteBase>[
+///         GoRoute(
+///           /// The screen to display as the root in the first tab of the
+///           /// bottom navigation bar.
+///           path: '/a',
+///           builder: (BuildContext context, GoRouterState state) =>
+///               const RootScreen(label: 'A', detailsPath: '/a/details'),
+///           routes: <RouteBase>[
+///             /// Will cover screen A but not the bottom navigation bar
+///             GoRoute(
+///               path: 'details',
+///               builder: (BuildContext context, GoRouterState state) =>
+///                   const DetailsScreen(label: 'A'),
+///             ),
+///           ],
+///         ),
+///         GoRoute(
+///           /// The screen to display as the root in the second tab of the
+///           /// bottom navigation bar.
+///           path: '/b',
+///           builder: (BuildContext context, GoRouterState state) =>
+///               const RootScreen(label: 'B', detailsPath: '/b/details'),
+///           routes: <RouteBase>[
+///             /// Will cover screen B but not the bottom navigation bar
+///             GoRoute(
+///               path: 'details',
+///               builder: (BuildContext context, GoRouterState state) =>
+///                   const DetailsScreen(label: 'B'),
+///             ),
+///           ],
+///         ),
+///       ],
+///       branches: <StatefulShellBranch>[
+///         StatefulShellBranch(rootLocation: '/a'),
+///         StatefulShellBranch(rootLocation: '/b'),
+///       ],
+///       builder: (BuildContext context, GoRouterState state, Widget child) {
+///         return ScaffoldWithNavBar(body: child);
+///       },
+///     ),
+///   ],
+/// );
+/// ```
+///
+/// When the [Page] for this route needs to be customized, you need to pass a
+/// function for pageBuilder. Note that this page builder doesn't replace
+/// the builder function, but instead receives the stateful shell built by
+/// [StatefulShellRoute] (using the builder function) as input. In other words,
```

Comment:
ah is it so that we can wrap a StatefulNavigationShell in between the page and the child?

If so I suggest we give them (builder/pageBuilder) a different name, The constructor can take these parameter and wire it up to only use the pageBuilder in the super. That way the builder.dart doesn't need to know these to build the page.

Reply:
Yes, exactly, and thus the builder is always required. And the pageBuilder is just used for customising the Page for the already built (and wrapped) shell. At one point I called pageBuilder pageProvider instead, but after feedback I changed it back to pageBuilder. But it might be good with different names that would make the purpose of the builders clearer. Some possible names for `builder` might perhaps be `shellBuilder` or `shellWidgetBuilder`. 

Regarding wiring up to use only the pageBuilder in the constructor StatefulShellRoute, I'm not sure about how to accomplish this in a clean way (if I understood you correctly that is), since builder.dart contains a lot of logic internally for default pageBuilder implementation as well as building the StatefulNavigationShell and the Navigator etc. Information such as PopPageCallback, RouteMatchList, page stack etc would have to be passed to the pageBuilder in that case, which might not be optimal. So the question is what would be gained by doing this (again -maybe I misunderstood something)? But I'll have to look at this a bit closer.

Reply:
My general concerns are we should avoid leaking out the implementation of StatefulShellRoute, Ideally, classes outside this file, i.e RouteBuilder, should only use the API of GoRoute and ShellRouteBase.

Reply:
Sure, I guess I can understand that concern from one perspective on API and SRP etc, depending somewhat on the intended responsibility of different parts of GoRouter. However, a lot of the implementation details and behaviour related to GoRoute and ShellRoute are already delegated to most of the core classes of GoRouter. If you look at the primary responsibility of each class, it feels like RouteBuilder already handles most of the actual implementation details related to the building of widgets, pages and navigators, and thus would seem to be a good place to handle StatefulNavigationShell. To me it feels like the RouteBase classes are more like light-weight abstractions of a routes, whose primary purpose is to provide a clear and simple public API. But I may be wrong here of course, and might miss some part of the bigger picture. 

One downside though of moving implementation details to StatefulShellRoute is that it would add dependencies to internal classes and types in GoRouter, that would then leak out in the public API.

Reply:
For `builder` and `pageBuilder`, I am still a bit confused about the reason the builder need to be required. If only uses pageBuilder, couldn't the child be the Navigator wrapped in `StatefulNavigationShell`?

> a lot of the implementation details and behaviour related to GoRoute and ShellRoute are already delegated to most of the core classes of GoRouter

That is true, but we should not make it worse.

> is that it would add dependencies to internal classes and types in GoRouter,

I think you meant the class in stateful_navigation_shell.dart? we can move the class to route.dart?